[회원 관리] 사용자 정보 CRUD 기능 구현 및 개선 (#2, #4, #5)

kebin0591 opened the pull request:
## 개요
회원 관리 기능의 기본적인 CRUD 작업을 구현하고 개선하는 내용을 포함하고 있습니다. 주요 변경 사항으로는 사용자 정보 수정 기능 개선, 로깅 강화, 프론트엔드 상호작용 개선, 사용자 목록 필터링 및 검색 기능 추가, UI 개선, 그리고 로그인 기능 개선이 있습니다.

## 주요 변경 사항
1. 사용자 관리 기본 구조 구현
   - UserMapper 인터페이스 및 UserDTO 모델 클래스 생성
   - UserDAO 인터페이스 및 UserDAOImpl 클래스 구현
   - UserService 인터페이스 및 UserServiceImpl 클래스 구현
   - UserController 클래스 추가

2. 데이터베이스 연동
   - UserMapper.xml에 CRUD 관련 SQL 쿼리 구현
   - application.yml에 데이터베이스 설정 추가

3. 사용자 CRUD 기능 구현
   - 사용자 조회, 추가, 수정, 삭제 기능 개발
   - 사용자 로그인 기능 구현

4. 프론트엔드 개발
   - 사용자 목록, 수정, 로그인, 등록 페이지 템플릿 추가
   - 메인 페이지(index.html) 임시 수정

5. 코드 품질 개선
   - JavaDoc 및 설명 주석 추가로 문서화 강화
   - null 체크 로직 개선 및 예외 처리 강화
   - ResultMap 도입으로 ORM 매핑 개선

6. 로깅 시스템 구축
   - UserController, UserServiceImpl에 상세 로깅 추가
   - application.yml에서 로깅 레벨 조정

7. 보안 및 데이터 무결성 강화
   - 사용자 랭크 수정 방지 로직 추가
   - 신규 사용자 등록 시 기본 랭크 설정

8. 성능 및 사용자 경험 개선
   - 프론트엔드에서 사용자 수정 링크 클릭 시 AJAX 처리 추가
   - 사용자 목록 및 수정 페이지 UI/UX 개선

9. 테스트 및 디버깅
   - UserMapperTest 작성 (추가 개선 예정)
   - 각 기능에 대한 테스트 및 버그 수정 수행

(0624 01시 내용 추가)

10. 사용자 목록 필터링 및 검색 기능 추가
    - UserController에 필터링 파라미터 추가 (userId, userName, userRank, startDate, endDate)
    - UserMapper에 selectFilteredUsers 메소드 추가
    - UserService와 UserServiceImpl에 getFilteredUsers 메소드 구현
    - UserMapper.xml에 selectFilteredUsers 쿼리 추가
    - list.html 템플릿에 검색 폼 추가 및 UI 개선

11. 사용자 검색 기능의 날짜 처리 개선 및 로깅 강화
    - UserController의 날짜 형식 패턴을 'yyyy-MM-dd'로 변경
    - UserServiceImpl에서 날짜 파라미터 처리 방식 개선
    - UserMapper.xml의 날짜 비교 쿼리 수정으로 정확성 향상
    - 디버그 및 정보 로깅 추가로 문제 해결 용이성 개선

12. 회원 관리 시스템 UI 개선 및 테스트 기능 추가
    - index.html: 메인 페이지 디자인 개선 및 테스트 기능 버튼 추가
    - user/list.html: 회원 목록 페이지 스타일 개선 및 테스트 기능 확장
    - user/register.html: 회원 등록 페이지 UI 개선 및 테스트 기능 구현

13. 로그인 기능 개선 및 UI 업데이트
    - UserController에 HttpSession을 추가하여 로그인 상태 관리
    - UserServiceImpl에 로그인 시도 로깅 추가
    - 로그인 페이지 UI 개선 및 테스트 기능 추가
    - 에러 메시지 한글화 및 예외 처리 개선
    - MyBatis 로깅 설정 변경

## 테스트 및 디버깅

모든 기능에 대한 테스트를 수행하고, 발견된 모든 버그를 성공적으로 수정했습니다.

- 테스트 완료: 사용자 정보 조회, 수정, 삭제 기능, 필터링 및 검색 기능
- 디버깅 완료: 랭크 수정 방지, 로깅 오류, AJAX 요청 처리, 날짜 처리 관련 이슈

## 관련 이슈
- #2
- #4
- #5

## 리뷰 요청사항
- 코드의 전반적인 구조와 가독성
- 예외 처리의 적절성
- 로깅 레벨 및 내용의 적절성
- 프론트엔드와 백엔드의 연동 부분
- 새로 추가된 필터링 및 검색 기능의 구현 방식

리뷰 부탁드려용용~~

## 추가 노트
이 PR은 현재 진행 중인 회원 관리 기능 구현 스프린트의 일부입니다. 각 메서드에 세부적인 비즈니스 로직 추가 예정이고, 프론트를 개발하면서 필요한 기능이 있으면 추가수정 예정입니다. 또한, 템플릿이 확정되면 프론트엔드 작업을 다시 진행할 수 있으며, 그에 따라 추가 수정이 있을 수 있습니다.

kebin0591 committed:
feat(user): 사용자 관리 기능 추가

- UserMapper 인터페이스 생성
- UserDTO 모델 클래스 구현
- UserMapper.xml 파일에 getAllUsers 쿼리 추가
- UserMapperTest 작성하여 getAllUsers 기능 테스트(에러나서 기능 더 추가한 후 수정 예정)

kebin0591 committed:
feat(user): 사용자 관리 기능 구현

- UserDAO 인터페이스 및 UserDAOImpl 클래스 추가
- UserMapper 인터페이스에 CRUD 메서드 추가
- UserMapper.xml에 SQL 쿼리 구현
- 사용자 조회, 추가, 수정, 삭제 및 로그인 기능 구현

kebin0591 committed:
feat(user): 사용자 서비스 레이어 구현

- UserService 인터페이스 추가
- UserServiceImpl 클래스로 UserService 구현
- 사용자 CRUD 및 로그인 기능 추가

향후 각 메서드에 비즈니스 로직 추가 예정

kebin0591 committed:
feat(user): 사용자 관리 기능 구현 및 문서화

- UserController 클래스 추가: 사용자 CRUD 및 로그인 기능 구현
- UserMapper 인터페이스에 JavaDoc 주석 추가
- UserService 및 UserDAO 인터페이스에 설명 주석 추가
- UserServiceImpl 및 UserDAOImpl 클래스에 설명 주석 추가

kebin0591 committed:
feat(user): 사용자 관리 기능 개선 및 로깅 추가

- UserController에 로깅 추가 및 예외 처리 개선
- UserServiceImpl에 상세 로깅 추가
- 사용자 목록, 수정, 로그인, 등록 페이지 템플릿 추가
- application.yml에 데이터베이스 설정 및 로깅 레벨 조정
- index.html을 메인 페이지 테스트로 임시 수정

kebin0591 committed:
refactor(user): 사용자 관리 기능 개선 및 버그 수정

- UserController에서 null 체크 로직 개선
- UserServiceImpl에서 신규 사용자 등록 시 기본 랭크 설정
- UserMapper.xml에서 ResultMap 도입으로 ORM 매핑 개선
- 사용자 목록 및 수정 테스트 페이지 UI/UX 개선
- 애플리케이션 이름 변경 및 로깅 레벨 조정

kebin0591 committed:
feat(user): 사용자 정보 수정 기능 개선 및 로깅 강화

- UserController에서 사용자 정보 수정 로직 개선
- 사용자 랭크 수정 방지 로직 추가(키 제약 위반)
- UserDAO, UserService에 상세 로깅 추가
- UserMapper.xml에서 모든 사용자 필드 매핑
- 프론트엔드에서 사용자 수정 링크 클릭 시 AJAX 처리 추가
- application.yml에서 로깅 레벨 조정

kebin0591 committed:
feat(user): 사용자 목록 필터링 및 검색 기능 추가

- UserController에 필터링 파라미터 추가 (userId, userName, userRank, startDate, endDate)
- UserMapper에 selectFilteredUsers 메소드 추가
- UserService와 UserServiceImpl에 getFilteredUsers 메소드 구현
- UserMapper.xml에 selectFilteredUsers 쿼리 추가
- list.html 템플릿에 검색 폼 추가 및 UI 개선
- 현재 테스트 중에 있음.

kebin0591 committed:
fix(user): 사용자 검색 기능의 날짜 처리 개선 및 로깅 강화

- UserController의 날짜 형식 패턴을 'yyyy-MM-dd'로 변경
- UserServiceImpl에서 날짜 파라미터 처리 방식 개선
- UserMapper.xml의 날짜 비교 쿼리 수정으로 정확성 향상
- 디버그 및 정보 로깅 추가로 문제 해결 용이성 개선
- 모든 변경 사항에 대한 기능 테스트 완료 및 정상 동작 확인

kebin0591 committed:
feat(ui): 회원 관리 시스템 UI 개선 및 테스트 기능 추가

- index.html: 메인 페이지 디자인 개선 및 테스트 기능 버튼 추가
- user/list.html: 회원 목록 페이지 스타일 개선 및 테스트 기능 확장
- user/register.html: 회원 등록 페이지 UI 개선 및 테스트 기능 구현
- UserServiceImpl.java: 코드 포맷팅 수정

테스트 페이지임을 명시하고, 사용자 경험을 개선하기 위해 UI를 전반적으로
업데이트했습니다.

kebin0591 committed:
feat(auth): 로그인 기능 개선 및 UI 업데이트

- UserController에 HttpSession을 추가하여 로그인 상태 관리
- UserServiceImpl에 로그인 시도 로깅 추가
- 로그인 페이지 UI 개선 및 테스트 기능 추가
- 에러 메시지 한글화 및 예외 처리 개선
- MyBatis 로깅 설정 변경

kebin0591 commented:
리뷰... 부탁....드려요,....

yujin4sth00 reviewed (approved):
확인했습니다.
UI와 회원 정보 수정 되는 건 확인 사진을 봤습니다.

DokaDev commented on src/main/java/com/nf404/devshop/controller/UserController.java:
```diff
@@ -0,0 +1,151 @@
+        return "redirect:/users/list";
+    }
+
+    @GetMapping("/delete/{userId}")
```
이건 Post요청으로 처리하시는게 어떨까 하는 의견입니다.

DokaDev replied:
url만으로도 path로 지정된 번호에 해당하는 사용자 정보를 지울 수 있게 되는 문제가 생길 수 있지 않을까 싶네요.

kebin0591 replied:
> url만으로도 path로 지정된 번호에 해당하는 사용자 정보를 지울 수 있게 되는 문제가 생길 수 있지 않을까 싶네요.

생각해보니 그런것 같습니다. 보완하겠습니다~

kebin0591 replied:
> url만으로도 path로 지정된 번호에 해당하는 사용자 정보를 지울 수 있게 되는 문제가 생길 수 있지 않을까 싶네요.

수정 완료해서 get에서 post로 바꿨습니다~

DokaDev commented on src/main/resources/application.yml:
url/username/password는 이전 값으로 바꿔주세요. 실 인증정보가 노출되지 않는 것이 바람직해 보입니다.

kebin0591 replied:
> url/username/password는 이전 값으로 바꿔주세요. 실 인증정보가 노출되지 않는 것이 바람직해 보입니다.

조금더 자세히 설명해주실수 있나용?

sanghyeonJwa commented:
혹시 UserDAO 와 UserMapper의 차이가 뭔가요?

sanghyeonJwa commented on src/main/java/com/nf404/devshop/dao/UserDAO.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * 사용자 데이터 액세스 객체(DAO)를 정의합니다.
+ * 이 인터페이스는 사용자 데이터에 대한 CRUD(Create, Read, Update, Delete) 작업을 제공.
+ * 사용자 데이터는 UserDTO 객체로 표현됩니다.
+ */
+package com.nf404.devshop.dao;
+
+import com.nf404.devshop.model.UserDTO;
+import java.util.List;
+
+public interface UserDAO {
+    // 모든 사용자 조회
+    List<UserDTO> getAllUsers();
+
+    // 사용자 ID로 사용자 조회
+    UserDTO getUserById(String userId);
+
+    // 새 사용자 등록
+    void insertUser(UserDTO user);
+
+    // 사용자 정보 업데이트
+    void updateUser(UserDTO user);
+
+    // 사용자 삭제
+    void deleteUser(String userId);
+
+    // 사용자 로그인 (ID와 비밀번호로 사용자 조회)
+    UserDTO loginUser(String userId, String userPw);
+}
```
이 부분과 UserMapper와의 차이가 궁금합니다!

kebin0591 replied:
> 이 부분과 UserMapper와의 차이가 궁금합니다!

매퍼는 mybatis용으로 만들었고 쿼리 작성용이거 DAO는 더 높은(?) 추상화를 제공해야해서 저렇게 나눴던거같습니다! 그리고 DAO에 대한 구현은 impl에서 해서 저렇게 짰던 것 같습니다.

sanghyeonJwa commented on src/main/java/com/nf404/devshop/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+    }
+
+
+    @Override
+    public void registerUser(UserDTO user) {
+        // 비즈니스 로직 추가 예정.
+        // 예: 비밀번호 암호화, 유효성 검사 등..
+        // 기본 랭크 설정 (예: 1)
+        user.setUserRank(1);
+        userDAO.insertUser(user);
+    }
+
+    @Override
+    public void updateUser(UserDTO user) {
+        // 비즈니스 로직 추가 예정.
+        // 예: 변경된 필드만 업데이트하는 로직 등
+        userDAO.updateUser(user);
+    }
+
+    @Override
+    public void deleteUser(String userId) {
+        userDAO.deleteUser(userId);
+    }
```
트랜잭션 어노테이션을 붙여야 하지 않을까 싶습니다..!

kebin0591 replied:
> 트랜잭션 어노테이션을 붙여야 하지 않을까 싶습니다..!

오 이따 자세히 알려주시죠

kebin0591 replied:
@sanghyeonJwa 서비스 부분에 CUD 관련하여 트랙잭션 어노테이션으로 수정 완료하였습니다! 
R부분은 Transactional(readOnly = true)를 이용하여 읽기 전용으로만 했습니다~!

sanghyeonJwa commented:
구현하신 부분 도메인 (member 혹은 user) 패키지 안에 넣어주시면 감사하겠습니다!

DokaDev commented:
아키텍처 간 의존성 주입 형태가 다소 부적절하게 사용된 부분이 보입니다. Service 및 Dao/Mapper단의 의존성주입 관계 확인 후 조정부탁드립니다. Dao/Mapper을 둘로 나눌 필요가 굳이 없다면 한 개념(sugg. 매퍼로...)으로 통일해주시면 좋겠습니다.

zZerorok commented:
꼼꼼하게 작성하셨네요. 
보고 참고할게요~~

kebin0591 commented:
> 꼼꼼하게 작성하셨네요. 
> 보고 참고할게요~~

어어.... 퍼가실때 출처 필수입니다~

kebin0591 committed:
docs(user): UserService 인터페이스에 설명 추가 및 트랜잭션 설정
- UserService 인터페이스의 각 메서드에 상세한 JavaDoc 주석 추가
- @transactional 어노테이션을 사용하여 트랜잭션 관리 설정

kebin0591 committed:
refactor(users): 회원 관리 파일 패키지 생성 및 병함

- 회원 관련 컨트롤러, 서비스 등 주요 회원 관리 기능 관련한 파일들을 user 패키지로 리팩터링 작업을 했습니다.

kebin0591 committed:
feat(user): 사용자 관리 기능 개선 및 보안 강화

- 사용자 삭제 메서드를 GET에서 POST로 변경하여 보안 강화

kebin0591 committed:
refactor: fixed application.yml

kebin0591 committed:
refactor(user): 사용하지 않는 UserDAO와 관련 클래스 제거

- 더 이상 필요하지 않은 UserDAO 인터페이스와 그 구현 클래스를 제거했습니다.
- UserServiceImpl 클래스에서 사용하지 않는 UserDAO 임포트를 제거했습니다.
- UserMapper를 직접 사용하도록 UserService 구현을 단순화했습니다.
